README.md
# AKVQ

Azure Key Vault Quick

(pronounced: A Quick (a /kwɪk/))

Get secrets from azure key vaults quick and easy


docs/changelog.rst
=========
Changelog
=========

v0.0.5
======

- add __version__
- better printout

v0.0.4
======

- renames akv secret names from dash to underscore
- working InteractiveBrowser and DeviceCode login
- support for setting vault_uri in git config
- better printout

v0.0.3
======

- fix entrypoint
- fix project structure
- add links to pyproject.toml

v0.0.2
======

- add entrypoint
- removed unused imports
- mitigated circular import

v0.0.1
======

- init version

src/akvq.py
import logging
import os
import sys
from pathlib import Path

from azure.core.exceptions import ClientAuthenticationError
from azure.identity import DefaultAzureCredential, TokenCachePersistenceOptions
from azure.keyvault.secrets import SecretClient


__doc__ = """

get all secrets and only set ENV on keys that start with "akvq-"
use alwyas UPPERCASE for ENV
SWAP dash to underscores, akv only allow (link to spec)

See :manpage:`bash#name`

akvq-HASS -> HASS (as env)
hello -> (skip)

For now I am skipping to implemt auth cache because we are saveing it to disk
https://github.com/Azure/azure-sdk-for-python/issues/11134

source ~/.akvq.sh
Follow for bash: https://stackoverflow.com/q/10735574

Uess the same as for ssh keys

:doc:`azure.keyvault.secrets`

:doc:`azure-keyvault-secrets:index`
"""


def main():
    """
    .. todo::

        - save secrets to file
        - clear file
        - add config list like poetry
        - env for vault_url
        - support to add to vault

    :class:`azure.identity.DefaultAzureCredential`

    :class:`azure.identity.TokenCachePersistenceOptions`

    `azure-identity logging
    <https://azuresdkdocs.blob.core.windows.net/$web/python/azure-identity/latest/index.html#logging>`_

    """
    # cache_options = TokenCachePersistenceOptions()

    # # From: https://stackoverflow.com/a/65893640
    # azure_identity_logger = logging.getLogger("azure.identity")

    # # Critical is to hide the error from azure
    # azure_identity_logger.setLevel(logging.CRITICAL)

    import logger
    log = logger.config("akvq")

    try:
        credential = DefaultAzureCredential(
            # cache_persistence_options=cache_options,
            exclude_interactive_browser_credential=False,
            logging_enable=True,
        )

        secret_client = SecretClient(
            vault_url=os.environ.get("AKVQ_VAULT_URI"),
            credential=credential,
        )

        # secret = secret_client.get_secret("hello")

        # print(f"secret value from keyvault is: '{secret.value}'")

        aprefix = "akvq-"
        secrets = []
        secret_properties = secret_client.list_properties_of_secrets()
        for secret_property in secret_properties:
            # the list doesn't include values or versions of the secrets
            if secret_property.name.startswith(aprefix):
                secrets.append(secret_property.name)

        akfilename = Path.home() / ".akvq.sh"

        with open(akfilename, "w") as akfile:
            # Will override oldfile
            for secret in secrets:
                real_secret_name = secret.split(aprefix)[1]
                log.info(f"adding: {real_secret_name}")
                secret_value = secret_client.get_secret(secret).value
                akfile.write(f"export {real_secret_name}={secret_value}\n")

    except ClientAuthenticationError:
        pass

    # # specify a cache name to isolate the cache from other applications
    # TokenCachePersistenceOptions(name="akvq")


if __name__ == "__main__":
    main()


src/akvq/logger.py
import logging
import logging.config

__doc__ = """
Logger
======

Based on:
https://docs.python.org/3/howto/logging-cookbook.html#formatting-times-using-utc-gmt-via-configuration

https://realpython.com/lessons/logger-dictionary/

The Galaxy Project has the same base:
https://docs.galaxyproject.org/en/release_20.05/admin/config_logging.html

https://docs.python.org/3/library/logging.html#logrecord-attributes
"""

from .settings import LOGGING

logging.getLogger(__name__).addHandler(logging.NullHandler())


def config(log_name):
    logging.config.dictConfig(LOGGING)
    log = logging.getLogger(log_name)
    self_log = logging.getLogger(__name__)
    self_log.debug(f"Loading log config: {LOGGING}")
    return log


def getLevelName(log_level):
    """Wapper for logging.getLevelName
    for skipping to import logging"""
    return logging.getLevelName(log_level)


src/akvq/settings.py
__doc__ = """
https://docs.djangoproject.com/en/4.0/topics/logging/

https://stackoverflow.com/a/33186558

"""
import logging


class UTCFormatter(logging.Formatter):
    # this formatter in the file to mitigate circular import
    converter = logging.time.gmtime


#
# START of LOGGING SETTINGS
#

DEBUG = "DEBUG"
INFO = "INFO"
WARNING = "WARNING"
ERROR = "ERROR"
CRITICAL = "CRITICAL"

LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "utc": {
            "()": UTCFormatter,
            "format": "[%(asctime)s.%(msecs)03dZ] %(levelname)s : %(name)s : %(message)s",
            "datefmt": "%Y-%m-%dT%H:%M:%S",
        }
    },
    "handlers": {"console": {"class": "logging.StreamHandler", "formatter": "utc"}},
    #
    # root is default logger
    "root": {"handlers": ["console"], "level": WARNING},
    #
    "loggers": {
        #
        # Define logger
        #
        # Hide failed azure login
        "azure.identity": {
            "level": CRITICAL,
        },
        # "azure.core": {
        #     "level": CRITICAL,
        # },
        "akvq": {
            "level": INFO,
        },
    },
}


src/akvq/__main__.py
import os
import subprocess
from pathlib import Path

from azure.core.exceptions import ClientAuthenticationError
from azure.identity import (
    ChainedTokenCredential,
    DeviceCodeCredential,
    InteractiveBrowserCredential,
)
from azure.keyvault.secrets import SecretClient

from . import logger

__doc__ = """

get all secrets and only set ENV on keys that start with "akvq-"
use alwyas UPPERCASE for ENV
SWAP dash to underscores, akv only allow (link to spec)

See :manpage:`bash#name`

akvq-HASS -> HASS (as env)
hello -> (skip)
akvq-DEMO-VAR -> DEMO_VAR (as env)

For now I am skipping to implemt auth cache because we are saveing it to disk
https://github.com/Azure/azure-sdk-for-python/issues/11134

source ~/.akvq.sh
Follow for bash: https://stackoverflow.com/q/10735574

Uess the same as for ssh keys

:doc:`azure.keyvault.secrets`

:doc:`azure-keyvault-secrets:index`
"""

# start main logger to get all settings
log = logger.config("akvq")

__version__ = "0.0.5"


def get_secrets() -> dict:
    """Get all AKVQ secrets from AKV, Expects secret_client"""
    log.critical("HELLOOOO")


def vault_uri() -> str:
    """
    Will check in order and return on match:

    - ENV
    - git config akvq.url

    Returns:
        str: url of azure key vault

    """
    uri = os.environ.get("AKVQ_VAULT_URI")
    if uri:
        log.debug("Got from ENV")

    if not uri:
        try:
            uri = (
                subprocess.check_output(["git", "config", "akvq.uri"]).decode().strip()
            )
            log.debug("Got from git config")
        except subprocess.CalledProcessError as exit_code:
            log.debug(exit_code)

    return uri


def main() -> int:
    """
    .. todo::

        - delete file on logout
        - add support for args, logout
        - support to add to vault
        - auto find vault based on tag
        - add support for windows

    """
    akvq_vault_uri = vault_uri()
    if not akvq_vault_uri:
        log.critical("Missing vault uri")
        return 1

    interactive_credential = InteractiveBrowserCredential()
    device_credential = DeviceCodeCredential()

    print(
        """A web browser has been opened at https://login.microsoftonline.com/organizations/oauth2/v2.0/authorize.
Please continue the login in the web browser.

If no web browser is available or if the web browser fails to open,
use the device code login."""
    )
    # https://azuresdkdocs.blob.core.windows.net/$web/python/azure-identity/latest/index.html#define-a-custom-authentication-flow-with-chainedtokencredential
    credential_chain = ChainedTokenCredential(interactive_credential, device_credential)

    secret_client = SecretClient(
        vault_url=vault_uri(), credential=credential_chain, logging_enable=True
    )

    aprefix = "akvq-"
    secrets = []
    secret_properties = secret_client.list_properties_of_secrets()
    for secret_property in secret_properties:
        # the list doesn't include values or versions of the secrets
        if secret_property.name.startswith(aprefix):
            secrets.append(secret_property.name)

    akfilename = Path.home() / ".akvq.env"
    print("Login successful")

    with open(akfilename, "w") as akfile:
        # Will override oldfile
        for secret in secrets:
            real_secret_name = secret.split(aprefix)[1]

            # AKV names are only allowed to contain alphanumeric characters and dashes
            # ENV names don't allow dash so a swap works good
            real_secret_name = real_secret_name.replace("-", "_")

            # log.info(f"adding: {real_secret_name}")
            print(f"adding: {real_secret_name}")
            secret_value = secret_client.get_secret(secret).value

            akfile.write(f"export {real_secret_name}={secret_value}\n")

    print("You might need to restart your terminal to source the new ENV")
    print(f"or run: source {akfilename}")


if __name__ == "__main__":
    main()
